chanfitness/chaneventstore.go

```diff
@@ -41,80 +42,77 @@ type ChannelEventStore struct {
 
 	wg sync.WaitGroup
 }
 
+// FlapCountFlushRate determines how often we write peer total flap count to
+// disk.
+var FlapCountFlushRate = time.Hour
+
 // Config provides the event store with functions required to monitor channel
 // activity. All elements of the config must be non-nil for the event store to
 // operate.
 type Config struct {
-	// SubscribeChannelEvents provides a subscription client which provides a
-	// stream of channel events.
+	// SubscribeChannelEvents provides a subscription client which provides
+	// a stream of channel events.
 	SubscribeChannelEvents func() (*subscribe.Client, error)
 
 	// SubscribePeerEvents provides a subscription client which provides a
 	// stream of peer online/offline events.
 	SubscribePeerEvents func() (*subscribe.Client, error)
 
-	// GetOpenChannels provides a list of existing open channels which is used
-	// to populate the ChannelEventStore with a set of channels on startup.
+	// GetOpenChannels provides a list of existing open channels which is
+	// used to populate the ChannelEventStore with a set of channels on
+	// startup.
 	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
-}
 
-// lifespanRequest contains the channel ID required to query the store for a
-// channel's lifespan and a blocking response channel on which the result is
-// sent.
-type lifespanRequest struct {
-	channelPoint wire.OutPoint
-	responseChan chan lifespanResponse
-}
+	// WriteFlapCount records our flap count for a peer on disk.
+	WriteFlapCount func(peer route.Vertex, flapCount uint32) error
+
+	// ReadFlapCount gets the flap count for a peer on disk.
+	ReadFlapCount func(peer route.Vertex) (uint32, error)
 
-// lifespanResponse contains the response to a lifespanRequest and an error if
-// one occurred.
-type lifespanResponse struct {
-	start time.Time
-	end   time.Time
-	err   error
+	// FlapCountTicker is a ticker which controls how often we flush our
+	// peer's flap count to disk.
+	FlapCountTicker ticker.Ticker
 }
 
-// uptimeRequest contains the parameters required to query the store for a
-// channel's uptime and a blocking response channel on which the result is sent.
-type uptimeRequest struct {
+type channelInfoRequest struct {
 	channelPoint wire.OutPoint
-	startTime    time.Time
-	endTime      time.Time
-	responseChan chan uptimeResponse
+	start        time.Time
+	end          time.Time
+	responseChan chan channelInfoResponse
 }
 
-// uptimeResponse contains the response to an uptimeRequest and an error if one
-// occurred.
-type uptimeResponse struct {
-	uptime time.Duration
-	err    error
+type channelInfoResponse struct {
+	info *ChannelInfo
+	err  error
 }
 
 // NewChannelEventStore initializes an event store with the config provided.
 // Note that this function does not start the main event loop, Start() must be
 // called.
-func NewChannelEventStore(config *Config) *ChannelEventStore {
+func NewChannelEventStore(config *Config,
+	now func() time.Time) *ChannelEventStore {
```

Comment:
Why did you decide to not put the `now` func inside the config? I'm not opposed but I'm curious as to your reasoning. I think it's cleaner to modify the config but there might be other reasons to *not* put it in the config.

Reply:
At the time, it didn't feel like "config", but no strong reasoning. Moved into `Config` and gone for `clock.Clock` rather than just a now function, more flexible that way.